build_zng.rs

```diff
@@ -1,58 +1,691 @@
+            const NEON: &[&str] = &["-mfpu=neon"];
+
+            let flags = if is_aarch64 { NEON_64 } else { NEON };
+
+            if ctx.enable(TargetFeature {
+                check: "neon",
+                defines: &["ARM_NEON"],
+                files: &[
+                    "adler32_neon",
+                    "chunkset_neon",
+                    "compare256_neon",
+                    "slide_hash_neon",
+                ],
+                flags,
+                msvc_flags: &[],
+            }) {
+                if msvc {
+                    ctx.append.defines.insert("__ARM_NEON__");
+                }
+
+                ctx.compile_check(
+                    "arm",
+                    TargetFeature {
+                        check: "ld4",
+                        defines: &["ARM_NEON_HASLD4"],
+                        files: &[],
+                        flags,
+                        msvc_flags: &[],
+                    },
+                );
+            }
         }
-        "zlibstatic"
+        _ => {
+            // TODO: powerpc, riscv, s390
+        }
+    }
+
+    let include = dst.join("include");
+
+    fs::create_dir_all(&include).unwrap();
+
+    let (zlib_h, mangle) = if compat {
+        fs::copy("src/zlib-ng/zconf.h.in", include.join("zconf.h")).unwrap();
+        ("zlib.h", "zlib_name_mangling.h")
     } else {
-        "z"
+        fs::copy("src/zlib-ng/zconf-ng.h.in", include.join("zconf-ng.h")).unwrap();
+        ("zlib-ng.h", "zlib_name_mangling-ng.h")
     };
-    println!(
-        "cargo:rustc-link-lib=static={}{}{}",
-        libname,
-        if compat { "" } else { "-ng" },
-        debug_suffix,
-    );
-    println!("cargo:root={}", install_dir.to_str().unwrap());
-    println!("cargo:include={}", includedir.to_str().unwrap());
+
+    fs::copy(
+        "src/zlib-ng/zlib_name_mangling.h.empty",
+        include.join(mangle),
+    )
+    .unwrap();
+
+    let version = {
+        let mut version = None;
+        strip_symbol_prefix(
+            Path::new(&format!("src/zlib-ng/{zlib_h}.in")),
+            &include.join(zlib_h),
+            |line| {
+                if line.contains("ZLIBNG_VERSION") && line.contains("#define") {
+                    version = Some(line.split('"').nth(1).unwrap().to_owned());
+                }
+            },
+        );
+
+        version.expect("failed to detect ZLIBNG_VERSION")
+    };
+
+    cfg.include(include).include("src/zlib-ng");
+    cfg.compile("z");
+
+    fs::create_dir_all(lib.join("pkgconfig")).unwrap();
+    fs::write(
+        lib.join("pkgconfig/zlib.pc"),
+        fs::read_to_string("src/zlib-ng/zlib.pc.in")
+            .unwrap()
+            .replace("@prefix@", dst.to_str().unwrap())
+            .replace("@includedir@", "${prefix}/include")
+            .replace("@libdir@", "${prefix}/lib")
+            .replace("@VERSION@", &version),
+    )
+    .unwrap();
+
+    let dst = dst.to_str().unwrap();
```

Comment:
`dst` could contain non utf-i characters:

```suggestion
    let dst = dst.display();
```

Reply:
This is a copy of what build.rs does.